driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -638,4 +720,310 @@ public Prune shouldPrune(final UsageTrackingInternalConnection usageTrackingConn
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace(format("Received opened connection [%s] to server %s",
+                                getId(availableConnection), serverId.getAddress()));
+                    }
+                }
+                return availableConnection;
+            } finally {
+                try {
+                    if (tryGetAvailable && availableConnection == null) {//the desired connection slot has not yet been removed
+                        removeLastDesiredConnectionSlot();
+                    }
+                } finally {
+                    lock.unlock();
+                }
+            }
+        }
+
+        private void releasePermit() {
+            lock();
+            try {
+                assert permits < maxPermits : permits;
+                permits++;
+                condition.signal();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        private void addLastDesiredConnectionSlot() {
+            desiredConnectionSlots.addLast(new MutableReference<>());
+        }
+
+        @Nullable
+        private PooledConnection tryGetAndRemoveFirstDesiredConnectionSlot() {
+            assert !desiredConnectionSlots.isEmpty();
+            PooledConnection result = desiredConnectionSlots.peekFirst().reference;
+            if (result != null) {
+                desiredConnectionSlots.removeFirst();
+            }
+            return result;
+        }
+
+        private void removeLastDesiredConnectionSlot() {
+            assert !desiredConnectionSlots.isEmpty();
+            PooledConnection connection = desiredConnectionSlots.removeLast().reference;
+            if (connection != null) {
+                connection.release();
+            }
+        }
+
+        boolean tryHandOver(final PooledConnection openConnection) {
+            lock();
+            try {
+                for (//iterate from first (head) to last (tail)
+                        MutableReference<PooledConnection> desiredConnectionSlot : desiredConnectionSlots) {
+                    if (desiredConnectionSlot.reference == null) {
+                        desiredConnectionSlot.reference = openConnection;
+                        condition.signal();
+                        if (LOGGER.isTraceEnabled()) {
+                            LOGGER.trace(format("Handed over opened connection [%s] to server %s",
+                                    getId(openConnection), serverId.getAddress()));
+                        }
+                        return true;
+                    }
+                }
+                return false;
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        @Nullable
+        private PooledConnection openAndReleasePermit(
+                final PooledConnection connection,
+                @Nullable final SingleResultCallback<InternalConnection> callback) {
```

Comment:
`asyncCallback` was initially called this way instead of being called `callback` to emphasize that it is completed asynchronously by a package-protected method. Private methods did not need this clarification, and the one method that had this parameter became private later. Anyway, the idea was senseless, I apologize for this naming.

Reply:
Don't stress - as you've seen from my own code being reviewed, being in the code is very different to reviewing the code. I find the picture is different, I work in close to solve the problem when coding, and its only through stepping back do you see the bigger picture.